File: node/src/runtime/provider-replay.ts
import type { LLMProvider, Message, ProviderDescriptor, ProviderReplay, RenderedContext, ReplayabilityAssessment, ToolCall } from "../types.js"
import type { SessionEvent } from "./session-log.js"

function sortObjectKeys(val: any): any {
  if (val === null || typeof val !== "object") {
    return val
  }
  if (Array.isArray(val)) {
    return val.map(sortObjectKeys)
  }
  const sortedKeys = Object.keys(val).sort()
  const result: Record<string, any> = {}
  for (const key of sortedKeys) {
    result[key] = sortObjectKeys(val[key])
  }
  return result
}

export function assistantReplayKey(message: Pick<Message, "content" | "toolCalls">): string {
  const toolCalls = (message.toolCalls ?? []).map(tc => {
    let normalizedArgs = tc.arguments
    try {
      const parsed = typeof tc.arguments === "string" ? JSON.parse(tc.arguments) : tc.arguments
      normalizedArgs = JSON.stringify(sortObjectKeys(parsed))
    } catch {
      // fallback
    }
    return {
      id: tc.id,
      name: tc.name,
      arguments: normalizedArgs,
    }
  })
  return JSON.stringify({
    content: message.content,
    toolCalls,
  })
}

/**
 * A stored replay may only be seeded into a provider speaking the same wire
 * protocol. On a cross-protocol fallback (provider A -> provider B) the
 * incompatible envelope is skipped so B re-serializes neutral context instead
 * of replaying A's protocol-specific shape.
 */
export function isReplayCompatibleWithProvider(
  replay: Partial<ProviderReplay>,
  descriptor: ProviderDescriptor | undefined,
): boolean {
  assertCanonicalReplay(replay)
  return !descriptor || replay.protocol === descriptor.protocol
}

const REPLAY_KEYS = new Set(["protocol", "provider", "model", "native_blocks", "reasoning_content", "reasoning_details", "native_message", "tool_calls"])

function assertCanonicalReplay(replay: Partial<ProviderReplay>): asserts replay is ProviderReplay {
  for (const key of Object.keys(replay)) if (!REPLAY_KEYS.has(key)) throw new Error(`provider replay has unknown field ${key}`)
  if (typeof replay.protocol !== "string" || replay.protocol.length === 0) throw new Error("provider replay protocol is required")
}

export function seedProviderReplayFromEvents(
  provider: LLMProvider,
  events: Array<{ event: SessionEvent }>,
): void {
  if (!provider.seedProviderReplay) return
  const descriptor = provider.descriptor?.()
  for (const { event } of events) {
    if (event.kind !== "llm_completed") continue
    const toolCalls = event.tool_calls ?? []
    const stored = event.provider_replay
    if (!stored || !isReplayCompatibleWithProvider(stored, descriptor)) continue
    provider.seedProviderReplay({ content: event.content, toolCalls }, stored)
  }
}

export function peekProviderReplay(
  provider: LLMProvider,
  content: string,
  toolCalls: ToolCall[],
): ProviderReplay | undefined {
  return provider.peekProviderReplay?.({ content, toolCalls })
}

/**
 * Pre-flight query for fallback routing: would `context` validate against
 * `provider` (with `extensions`) before the request is sent? Seed any persisted
 * replay (via `seedProviderReplayFromEvents`) first so the assessment reflects
 * what the provider can actually replay. Providers that do not implement
 * `assessReplayability` (no reasoning-replay requirement) are reported as ok.
 */
export function assessProviderReplayability(
  provider: LLMProvider,
  context: RenderedContext,
  extensions?: Record<string, unknown>,
): ReplayabilityAssessment {
  return provider.assessReplayability?.(context, extensions) ?? { ok: true, offendingCallIds: [] }
}

File: rust/src/runtime/provider_replay.rs
use deepstrike_core::runtime::session::{ProviderReplay, SessionEvent};
use deepstrike_core::types::message::ToolCall;

use super::session_log::SessionEntry;

pub fn seed_provider_replay_from_events(
    provider: &dyn crate::providers::LLMProvider,
    events: &[SessionEntry],
) {
    for entry in events {
        if let SessionEvent::LlmCompleted {
            message,
            provider_replay,
            ..
        } = &entry.event
        {
            if let Some(replay) = provider_replay {
                let content = message.content.as_text().unwrap_or("").to_string();
                provider.seed_provider_replay(&content, &message.tool_calls, replay);
            }
        }
    }
}

pub fn peek_provider_replay(
    provider: &dyn crate::providers::LLMProvider,
    content: &str,
    tool_calls: &[ToolCall],
) -> Option<ProviderReplay> {
    provider.peek_provider_replay(content, tool_calls)
}

pub fn assistant_replay_key(content: &str, tool_calls: &[ToolCall]) -> String {
    serde_json::json!({
        "content": content,
        "toolCalls": tool_calls.iter().map(|tc| {
            serde_json::json!({
                "id": tc.id.as_str(),
                "name": tc.name.as_str(),
                "arguments": tc.arguments.to_string(),
            })
        }).collect::<Vec<_>>(),
    })
    .to_string()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::providers::{LLMProvider, ProviderRunState, StreamEvent};
    use async_trait::async_trait;
    use deepstrike_core::context::renderer::RenderedContext;
    use deepstrike_core::runtime::session::SessionEvent;
    use deepstrike_core::types::message::{Content, Message, Role, ToolSchema};
    use futures::{Stream, stream};
    use std::sync::Mutex;

    #[derive(Default)]
    struct CapturingProvider {
        seeded: Mutex<Vec<(String, ProviderReplay)>>,
    }

    #[async_trait]
    impl LLMProvider for CapturingProvider {
        fn seed_provider_replay(
            &self,
            content: &str,
            _tool_calls: &[ToolCall],
            replay: &ProviderReplay,
        ) {
            self.seeded
                .lock()
                .unwrap()
                .push((content.to_owned(), replay.clone()));
        }

        async fn stream(
            &self,
            _context: &RenderedContext,
            _tools: &[ToolSchema],
            _extensions: Option<&serde_json::Value>,
            _state: Option<&ProviderRunState>,
        ) -> crate::Result<Box<dyn Stream<Item = crate::Result<StreamEvent>> + Send + Unpin>>
        {
            Ok(Box::new(Box::pin(stream::empty())))
        }
    }

    #[test]
    fn seeds_only_persisted_provider_replay_from_events() {
        let provider = CapturingProvider::default();
        let replay = ProviderReplay {
            native_blocks: None,
            reasoning_content: Some("trace".into()),
            extra: serde_json::Map::new(),
        };
        let entries = vec![
            SessionEntry {
                seq: 0,
                event: SessionEvent::LlmCompleted {
                    turn: 0,
                    message: Message {
                        role: Role::Assistant,
                        content: Content::Text("without replay".into()),
                        tool_calls: vec![],
                        token_count: None,
                    },
                    provider_replay: None,
                },
            },
            SessionEntry {
                seq: 1,
                event: SessionEvent::LlmCompleted {
                    turn: 1,
                    message: Message {
                        role: Role::Assistant,
                        content: Content::Text("with replay".into()),
                        tool_calls: vec![],
                        token_count: None,
                    },
                    provider_replay: Some(replay.clone()),
                },
            },
        ];

        seed_provider_replay_from_events(&provider, &entries);

        let seeded = provider.seeded.lock().unwrap();
        assert_eq!(seeded.len(), 1);
        assert_eq!(seeded[0].0, "with replay");
        assert_eq!(seeded[0].1, replay);
    }
}
